storage/src/merkle_storage.rs

```diff
@@ -2123,6 +1656,95 @@ mod tests {
+
+        storage.set(&vec!["a".to_string()], &vec![1u8, 2u8]);
+        assert_eq!(storage.dirmem(&key_a).unwrap(), false);
+        assert_eq!(storage.mem(&key_a).unwrap(), true);
+        storage.delete(&vec!["a".to_string()]);
+        assert_eq!(storage.dirmem(&key_a).unwrap(), false);
+        assert_eq!(storage.mem(&key_a).unwrap(), false);
+    }
+
+    #[test]
+    fn test_delete_whole_tree2() {
+        let db_name = "test_delete_whole_tree2";
+        clean_db(db_name);
+
+        let key_abcd: &ContextKey = &vec![
+            "a".to_string(),
+            "b".to_string(),
+            "c".to_string(),
+            "d".to_string(),
+        ];
+        let key_abc: &ContextKey = &vec!["a".to_string(), "b".to_string(), "c".to_string()];
+        let key_ab: &ContextKey = &vec!["a".to_string(), "b".to_string()];
+        let key_a: &ContextKey = &vec!["a".to_string()];
+
+        let cache = Cache::new_lru_cache(32 * 1024 * 1024).unwrap();
+        let mut storage = get_storage(db_name, &cache);
+
+        storage.set(key_abcd, &vec![1u8, 2u8]);
+
+        assert_eq!(storage.dirmem(&key_a).unwrap(), true);
+        assert_eq!(storage.dirmem(&key_ab).unwrap(), true);
+        assert_eq!(storage.dirmem(&key_abc).unwrap(), true);
+        assert_eq!(storage.mem(&key_abcd).unwrap(), true);
+
+        storage.delete(key_abcd);
+
+        assert_eq!(storage.dirmem(&key_a).unwrap(), false);
+        assert_eq!(storage.dirmem(&key_ab).unwrap(), false);
+        assert_eq!(storage.dirmem(&key_abc).unwrap(), false);
+        assert_eq!(storage.mem(&key_abcd).unwrap(), false);
+    }
+
+    #[test]
+    fn test_delete_subtree() {
+        let db_name = "test_delete_subtree";
+        clean_db(db_name);
+
+        let key_abcd: &ContextKey = &vec![
+            "a".to_string(),
+            "b".to_string(),
+            "c".to_string(),
+            "d".to_string(),
+        ];
+        let key_abc: &ContextKey = &vec!["a".to_string(), "b".to_string(), "c".to_string()];
+        let key_ab: &ContextKey = &vec!["a".to_string(), "b".to_string()];
+        let key_a: &ContextKey = &vec!["a".to_string()];
+
+        let key_ax: &ContextKey = &vec!["a".to_string(), "x".to_string()];
+
+        let cache = Cache::new_lru_cache(32 * 1024 * 1024).unwrap();
+        let mut storage = get_storage(db_name, &cache);
+
+        storage.set(key_abcd, &vec![1u8, 2u8]);
+        storage.set(key_ax, &vec![1u8, 2u8]);
+
+        assert_eq!(storage.dirmem(&key_a).unwrap(), true);
+        assert_eq!(storage.dirmem(&key_ab).unwrap(), true);
+        assert_eq!(storage.dirmem(&key_abc).unwrap(), true);
+        assert_eq!(storage.mem(&key_abcd).unwrap(), true);
+
+        assert_eq!(storage.mem(&key_ax).unwrap(), true);
+
+        storage.delete(key_abcd);
+
+        assert_eq!(storage.dirmem(&key_a).unwrap(), true);
+        assert_eq!(storage.mem(&key_ax).unwrap(), true);
+        assert_eq!(storage.dirmem(&key_ab).unwrap(), false);
```

Comment:
I am confused, shouldn't these be `true` ? `a/b/c/d` is being deleted, which should still keep the upper "directories" in place, or is there something I am missing?

Reply:
Oh, empty directories just shouldn't be there? I get it now.